### ISSUE TITLE
[2.0.3] options.plugins `{Function}` don't get applied (`webpack.config.js`)

### ISSUE THREAD

**Issue by Malenconiaprincep** [postcss/postcss-loader#218]
webpack.config.js

```
var webpack = require('webpack');
module.exports = {
  entry: __dirname + '/index.js',
  output: {
    publicPath: '/',
    filename: './bundle.js'
  },
  module: {
    loaders: [
      {
        test: /\.jsx?$/,
        exclude: /node_modules/,
        loader: 'babel-loader',
        query: {
          presets: ['es2015', 'stage-0', 'react']
        }
      },
      {
          test : /\.css$/,
          use: [
            'style-loader',
            {
              loader: 'css-loader',
              options: {
                importLoaders: 1 // 0 => no loaders (default); 1 => postcss-loader; 2 => postcss-loader, sass-loader
              }
            },
            {
              loader: 'postcss-loader',
              options: {
                plugins: function () {
                  return [
                    require('precss'),
                    require('autoprefixer')
                  ];
                }
              }
            }
          ]
      }
    ]
  }
};

```

my css 

App.css

```
@import 'config.css';

.title {
  display: flex;
  background: $blue;
  width: $column
}
```
config.css

```
$blue: #ff0000;
$column: 200px;
```

![image](https://cloud.githubusercontent.com/assets/6262739/25840353/1d79378c-34ce-11e7-92be-911f161ee13c.png)

package.json

![image](https://cloud.githubusercontent.com/assets/6262739/25840401/3fbc1f6c-34ce-11e7-8b83-8e2e5f4be200.png)

**Comment by michael-ciniawsky** [comment]
```diff
{
  loader: 'postcss-loader',
  options: {
    plugins: function () {
      return [
-        require('precss'),
+       require('precss')(),
-        require('autoprefixer')()
+       require('autoprefixer')()
      ];
   }
}
```

If I remember right we had issues with precss (Plugin Package), maybe try without

```
ERROR in ./~/css-loader?{...options}!./~/postcss-loader/lib?{} <= isEmpty
```

The options are not picked up and then `postcss-loader` tries to look for a `postcss.config.js`

Off topic, but please update webpack-dev-server to `v2.x` it's definitely partially incompatible with webpack >= `v2.0.0` 😛

**Comment by Malenconiaprincep** [comment]
Thanks for your advice
I upgrade webpack-dev-server to v2.4.5  and modify code , but still not work; I think my config loader postcss-loaders has plugins config that why it still tries to look for a ```postcss.config.js```

**Comment by michael-ciniawsky** [comment]
Please try `{ plugins: [ require('')(), require('')() ] }` && `{ plugins: () => [...plugins] }` aswell and did you remove `precss` ?

Off topic again, but please also remove `var webpack= require('webpack')` at the top of your `webpack.config.js`, it's unnecessary there :)

**Comment by michael-ciniawsky** [comment]
And... it shouldn't be needed anymore 'normally', but in webpack `< 2.2.1`, one needed to use an **identifier**, when using so called `Complex Options` (`{Function}`/`require()`), so try this aswell

**webpack.config.js**
```js
{
  loader: 'postcss-loader',
  options: {
    ident: 'postcss', // <= this line
    plugins: () => [...plugins]
  }
}
```

**Comment by Malenconiaprincep** [comment]
```
{
        loader: 'postcss-loader',
        options: {
               plugins: function () {
                    return [
                      require('autoprefixer')
                    ];
             }
        }
}
```
I remove precess , it can works , but I need to use @import colors.css variable , is plugin ```precss``` has problem in postcss-loader?

**Comment by michael-ciniawsky** [comment]
I did start to debug this and ` { plugins: () => [...plugins] }` 🛑  while `{ plugins: [...plugins] }` ✅ , (it's a bug in webpack), you can try again with `precss`, but for `@import` I recommend using `postcss-import` instead, some of the 'sassy' features of `precss` are better achieved with manually selected plugins aswell (lesser bloat) e.g `postcss-nested` for nesting selectors etc... 😛

**Comment by Malenconiaprincep** [comment]
Thanks for your advice

```
module.exports = {
  plugins: {
    'postcss-import': {},
    'autoprefixer': {},
    // Enjoy a familiar syntax with variables, mixins, conditionals, and other goodies.
    'precss':{},
  }
}
```

It's strange , I use postcss.config.js instead of webpack's config  , then I use postcss.config.js is work , but webpack's config is not work ~~~  

![image](https://cloud.githubusercontent.com/assets/6262739/25848272/300f13f8-34ec-11e7-8eaf-1ed194c4a717.png)

I follow your example ，but not work   😛

**Comment by michael-ciniawsky** [comment]
```js
{
  loader: 'postcss-loader'
  options: {
    plugins: [ // no function () => {} 
       require('plugin'),
       ...
    ]
  }
}
```

The `{ plugins: (loader) => [...plugins] }` bugs for an unknown reason atm, use `postcss.config.js` instead or an `{Array}` in `options.plugins` in the meantime

**Comment by Malenconiaprincep** [comment]
Thanks , use ``` {Array}``` in ```options.plugins``` is OK,  thanks u  😛

**Comment by michael-ciniawsky** [comment]
I keep this issue open for tracking porposes since a `{Function}` should normally work

**Comment by moimael** [comment]
Using an array lead to weird output in webpack compilation:
```
{"plugins":[null,{"version":"5.2.16","plugins":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],"postcssPlugin":"postcss-cssnext","postcssVersion":"5.2.16"},{"version":"5.2.16","plugins":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],"postcssPlugin":"cssnano","postcssVersion":"5.2.16"},null]}
```

**Comment by evilebottnawi** [comment]
@moimael In the near future (today) I will tackle this problem :+1:

**Comment by michael-ciniawsky** [comment]
@moimael But the CSS is getting minified? Locally for me that's the case, but there is bug somewhere in webpack (where the so called 'Complex Options' get applied (`{Function}/require()`))
Super annonying stuff, it's the third time this happens 👎

**Comment by moimael** [comment]
Yes minification seems to work. The output is just a bit weird.

**Comment by michael-ciniawsky** [comment]
Fixed && Released in [v2.0.5](https://github.com/postcss/postcss-loader/releases/tag/v2.0.5)

**Comment by zhe-he** [comment]
@michael-ciniawsky 
when i use Webpack code splitting feature.
`require.ensure(['./xxx.vue'], () => {
    resolve(require('./xxx.vue'))
  })
`
---    
xxx.vue
![af25a702-db6e-4966-ae7d-7f8cdd34288d](https://cloud.githubusercontent.com/assets/11306583/26776567/30883f78-4a0c-11e7-9cfe-4136015c3f39.png)

error: 
`ERROR in ./~/_css-loader@0.27.3@css-loader!./~/_postcss-loader@1.3.3@postcss-loader?{"config":{"path":"config/postcss.config.js"}}!./~/_sass-loader@6.0.5@sass-loader/lib/loader.js!./css/credentials.scss
Module build failed: TypeError: Path must be a string. Received { path: 'config/postcss.config.js' }`

---     
webpack.config.js .
``var loaders = [
    {loader: 'css-loader'},
    {
        loader: 'postcss-loader',
        options: {
            config: {
                path: 'config/postcss.config.js'
            }
        }
    }
];``    
``{
                test: /\.(scss|sass)$/,
                exclude:/node_modules/,
                use: ExtractTextPlugin.extract({
                    fallback: 'style-loader',
                    use: loaders.concat({loader: 'sass-loader'})
                })
            },``    
``{
                test: /\.vue$/,
                exclude:/node_modules/,
                use: [
                    {
                        loader: 'vue-loader',
                        options: {
                            extractCSS: true,
                            preserveWhitespace: false,
                            postcss: [ autoprefixer({browsers: ['last 9 versions']}) ],
                            loaders: {
                                'js': 'babel-loader?presets[]=es2015&plugins[]=transform-object-rest-spread'
                              
                            }
                        }
                    }
                ]
            },``    
---    
postcss.config.js.    
``const autoprefixer = require('autoprefixer')
module.exports = {
    plugins: [
        autoprefixer({ browsers: ['last 9 versions'], cascade: false })
    ]
}``    
---    
postcss-loader@2.0.5 .
vue-loader@11.3.4.
sass-loader@6.0.3.

**Comment by psirenny** [comment]
I think this issue might be back with webpack 3. `loaderUtils.getOptions()` seems to be discarding functions passed into it. I'm guessing because they are "non-serializable". See https://github.com/webpack/loader-utils/issues/56.

**Comment by michael-ciniawsky** [comment]
@psirenny I update the docs soon, I was reluctant tbh, bc this should be handle automatically since `webpack >= 2.2.1` you need to add an `ident: 'postcss'` to loader options, bc as you mentioned the options need to be serialized by webpack internally.

**webpack.config.js**
```js
{
   loader: 'postcss-loader'
   options: {
      ident: 'postcss',
      plugins: () => [ require('postcss-plugin')({...options}) ]
   }
}
```

**Comment by edwardfhsiao** [comment]
actually,
```js
ident: 'postcss',
```
fixed my problem.
My problem was that when I just start building using webpack3 watch mode, postcss-loader worked fine, however, if I got typos to have a syntax error in a js file, for example, I accidentally typed two commas, the js file broked, right? 
![image](https://user-images.githubusercontent.com/11728228/28101159-2b16507a-66f9-11e7-8e8e-182bdada7d89.png)

Right after I fixed the comma thing, postcss-loader immediately had this error and my css file broked.
"Module build failed: ModuleBuildError: Module build failed: TypeError: Cannot read property 'postcss' of null"
![image](https://user-images.githubusercontent.com/11728228/28101190-6a02af68-66f9-11e7-89df-f52f1022e959.png)

![image](https://user-images.githubusercontent.com/11728228/28101184-556e8144-66f9-11e7-9f4f-e42c8b11bc34.png)
